src/Components/Server/src/ComponentsHub.cs

```diff
@@ -42,32 +47,97 @@ public ComponentsHub(IServiceProvider services, ILogger<ComponentsHub> logger)
         }
 
+        // For unit testing
+        internal Task OnBeforeDeactivate { get; set; } = Task.CompletedTask;
+        internal Task OnBeforeActivate { get; set; } = Task.CompletedTask;
+
         /// <summary>
         /// Intended for framework use only. Applications should not call this method directly.
         /// </summary>
         public override async Task OnDisconnectedAsync(Exception exception)
         {
-            await CircuitHost.DisposeAsync();
+            var circuitHost = CircuitHost;
+            if (circuitHost == null)
+            {
+                return;
+            }
+
+            CircuitHost = null;
+            try
+            {
+                if (!await AcquireCircuitLock())
+                {
+                    // If we're unable to enter the critical section, clear this instance and throw.
+                    await circuitHost.DisposeAsync();
+
+                    throw new InvalidOperationException("Failed to acqiure a lock to de-activate the CircuitHost instance");
+                }
+
+                await OnBeforeDeactivate;
+
+                _circuitRegistry.Deactivate(circuitHost, Context.ConnectionId);
+            }
+            finally
+            {
+                CircuitRegistryLock.Release();
+            }
+
+            await circuitHost.OnConnectionDownAsync();
         }
 
         /// <summary>
         /// Intended for framework use only. Applications should not call this method directly.
         /// </summary>
-        public async Task StartCircuit(string uriAbsolute, string baseUriAbsolute)
+        public async Task<string> StartCircuit(string uriAbsolute, string baseUriAbsolute)
         {
-            var circuitHost = _circuitFactory.CreateCircuitHost(Context.GetHttpContext(), Clients.Caller);
+            var circuitHost = _circuitFactory.CreateCircuitHost(Context, Clients.Caller);
             circuitHost.UnhandledException += CircuitHost_UnhandledException;
-
-            var uriHelper = (RemoteUriHelper)circuitHost.Services.GetRequiredService<IUriHelper>();
-            uriHelper.Initialize(uriAbsolute, baseUriAbsolute);
+            circuitHost.RemoteUriHelper.Initialize(uriAbsolute, baseUriAbsolute);
 
             // If initialization fails, this will throw. The caller will fail if they try to call into any interop API.
             await circuitHost.InitializeAsync(Context.ConnectionAborted);
 
+            _circuitRegistry.Register(circuitHost);
+
             CircuitHost = circuitHost;
+
+            return circuitHost.CircuitId;
+        }
+
+        /// <summary>
+        /// Intended for framework use only. Applications should not call this method directly.
+        /// </summary>
+        public async Task<bool> ConnectCircuit(string circuitId)
+        {
+            CircuitHost circuitHost;
+            try
+            {
+                if (!await AcquireCircuitLock())
+                {
+                    return false;
+                }
```

Comment:
Is this **required** to connect to an existing circuit? ie - is the client smart enough to make a new one if this fails?

Reply:
Not as yet. The way I thought of this was you would want tombstoning if the client were to connect to a new circuit. We could connect to a completely new instance now, but you wouldn't have any of the state preserved which would be pretty frustrating as a user.